### PR TITLE
product custom gateway response on errors

### PR DESCRIPTION
**Pull request by eguzki**
Product CRD allows configuring custom gateway response on errors

```
apiVersion: capabilities.3scale.net/v1beta1
kind: Product
metadata:
  name: product1
spec:
  name: "OperatedProduct 1"
  backendUsages:
    backenda:
      path: /A
  deployment:
    apicastHosted:
      authentication:
        userkey:
          gatewayResponse:
            errorStatusAuthFailed: 500
            errorHeadersAuthFailed: "text/plain; charset=mycharset"
            errorAuthFailed: "My custom reponse body"
            errorStatusAuthMissing: 500
            errorHeadersAuthMissing: "text/plain; charset=mycharset"
            errorAuthMissing: "My custom reponse body"
            errorStatusNoMatch: 501
            errorHeadersNoMatch: "text/plain; charset=mycharset"
            errorNoMatch: "My custom reponse body"
            errorStatusLimitsExceeded: 502
            errorHeadersLimitsExceeded: "text/plain; charset=mycharset"
            errorLimitsExceeded: "My custom reponse body"
```

### CODE REVIEW

File: apis/capabilities/v1beta1/product_types.go [hunk 1]
```diff
@@ -224,7 +224,8 @@ type AppKeyAppIDAuthenticationSpec struct {
 	// +optional
 	Security *SecuritySpec `json:"security,omitempty"`
 
-	// TODO GATEWAY RESPONSE
+	// +optional
+	GatewayResponse *GatewayResponseSpec `json:"gatewayResponse,omitempty"`
```
Comment (miguelsorianod) [inline apis/capabilities/v1beta1/product_types.go:228]: Is this independent on how `ProductDeploymentSpec` is configured?

File: config/samples/capabilities_v1beta1_product_gateway_response.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+apiVersion: capabilities.3scale.net/v1beta1
```
Comment (miguelsorianod) [inline config/samples/capabilities_v1beta1_product_gateway_response.yaml:1]: We should add this new sample in the CRD tests

File: apis/capabilities/v1beta1/product_types.go (hunk 1 above)
Comment (miguelsorianod) [reply apis/capabilities/v1beta1/product_types.go:228]: I think this comment answers my question: https://github.com/3scale/3scale-operator/pull/528/files?file-filters%5B%5D=.go&file-filters%5B%5D=.md#diff-bb6a1b34101270566aae5073dcb3e50543c2058eea30c95cc18957f033dbb0b4R585

I understand then it is independent and both modes accept the same set of fields in System

**Review by miguelsorianod — CHANGES_REQUESTED** [review]
Looks good to me :+1: .

I left some minor comments

File: config/samples/capabilities_v1beta1_product_gateway_response.yaml (hunk 2 above)
Comment (eguzki) [reply config/samples/capabilities_v1beta1_product_gateway_response.yaml:1]: crd_validation_test.go automatically tests all samples based on file prefix name. It is already tested:

```
$ make test-crds
.....
=== RUN   TestSampleCustomResources/capabilities_v1beta1_product_gateway_response.yaml
--- PASS: TestSampleCustomResources (0.22s)
    .....
    --- PASS: TestSampleCustomResources/capabilities_v1beta1_product_gateway_response.yaml (0.00s)

```

Comment (miguelsorianod) [reply config/samples/capabilities_v1beta1_product_gateway_response.yaml:1]: :+1:

**Review by miguelsorianod — APPROVED** [review]

**Pull request merged**